cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go

```diff
@@ -70,43 +83,222 @@ func (o *checkCapacityProvClass) Provision(
+			scaleUpStatus, _ := status.UpdateScaleUpError(&status.ScaleUpStatus{}, errors.NewAutoscalerError(errors.InternalError, "error during ScaleUp: %s", err.Error()))
+			combinedStatus.Add(scaleUpStatus)
+		}
+
+		if scaleUpIsSuccessful && err == nil {
+			combinedStatus.Add(&status.ScaleUpStatus{Result: status.ScaleUpSuccessful})
+
+			if err := o.context.ClusterSnapshot.Commit(); err != nil {
+				scaleUpStatus, _ := status.UpdateScaleUpError(&status.ScaleUpStatus{}, errors.NewAutoscalerError(errors.InternalError, "error during ScaleUp: %s", err.Error()))
+				o.context.ClusterSnapshot.Revert()
+
+				combinedStatus.Add(scaleUpStatus)
+			}
+		} else {
+			combinedStatus.Add(&status.ScaleUpStatus{Result: status.ScaleUpNoOptionsAvailable})
+		}
+
+		provisioningRequestsProcessed++
+		if stopBatch := o.shouldStopBatchProcessing(provisioningRequestsProcessed, startTime); stopBatch {
+			break
+		}
+
+		// For batch processing, the next ProvisioningRequest's pods are fetched using the provisioningRequest pods injector.
+		// TODO: Refactor such that injector injects pods for multiple ProvisioningRequests at once. Pods should be segregated by ProvisioningRequest before being processed.
+		prPods, err = o.getNextPrPods()
+		if err != nil {
+			st, _ := status.UpdateScaleUpError(&status.ScaleUpStatus{}, errors.NewAutoscalerError(errors.InternalError, "error during ScaleUp: %s", err.Error()))
+			combinedStatus.Add(st)
+		}
 	}
-	return &status.ScaleUpStatus{Result: status.ScaleUpNoOptionsAvailable}, nil
+
+	return combinedStatus.Export()
 }
 
-// Assuming that all unschedulable pods comes from one ProvisioningRequest.
+// checkcapacity checks if there is capacity in the cluster for pods from a ProvisioningRequest. It persists the scheduling in cluster snapshot whenever all pods can be scheduled, and doesn't change the snapshot otherwise.
 func (o *checkCapacityProvClass) checkcapacity(unschedulablePods []*apiv1.Pod, provReq *provreqwrapper.ProvisioningRequest) (capacityAvailable bool, err error) {
 	capacityAvailable = true
-	st, _, err := o.injector.TrySchedulePods(o.context.ClusterSnapshot, unschedulablePods, scheduling.ScheduleAnywhere, true)
+
+	o.context.ClusterSnapshot.Fork()
+	st, _, err := o.schedulingSimulator.TrySchedulePods(o.context.ClusterSnapshot, unschedulablePods, scheduling.ScheduleAnywhere, true)
 	if len(st) < len(unschedulablePods) || err != nil {
 		conditions.AddOrUpdateCondition(provReq, v1.Provisioned, metav1.ConditionFalse, conditions.CapacityIsNotFoundReason, "Capacity is not found, CA will try to find it later.", metav1.Now())
 		capacityAvailable = false
+		o.context.ClusterSnapshot.Revert()
 	} else {
+		commitErr := o.context.ClusterSnapshot.Commit()
+		if commitErr != nil {
+			return false, fmt.Errorf("failed to commit cluster snapshot: %v", err)
+		}
+
 		conditions.AddOrUpdateCondition(provReq, v1.Provisioned, metav1.ConditionTrue, conditions.CapacityIsFoundReason, conditions.CapacityIsFoundMsg, metav1.Now())
 	}
+
```

Comment:
Hmm, shouldn't we also revert if the provreq update below fails? 

The following pattern might be useful:
```
func f() {
  var err error
  snapshot.Fork()
  defer func() {
    if err != nil {
      snapshot.Commit()
    } else {
      snapshot.Revert()
    }
  }
  ...
  rest of f() that sets err
}
```

Reply:
Refactored accordingly.

Reply:
Why not `Commit()` in the defer as well?

Reply:
If there is an error when trying to commit the cluster snapshot, we want to return immediately without adding the Provisioned=True condition to the ProvisioningRequest.